lib/RGBot.js

```diff
@@ -400,39 +769,142 @@ const RGBot = class {
+                            if (!this.#bot.heldItem || attackItem.name !== this.#bot.heldItem.name) {
+                                this.#log(`Equipping attack weapon: ${attackItem.name}`)
+                                await this.#bot.equip(attackItem, 'hand');
+                            }
+                        }
+                        this.#logDebug(`Attacking entity: ${entity.username || entity.name} using weapon: ${attackItem?.name}`)
+                        this.lastAttackTime = Date.now()
+                        this.lastAttackItem = attackItem
+                        this.#bot.attack(entity)
+                        return true
+                    } catch (ex) {
+                        this.#logDebug(`Failed to get near enough to entity to attack`);
+                    }
+                }
+            } catch (ex) {
+                this.#logWarn(`Error trying to attack entity: ${entity.username || entity.name}`, ex)
             }
         }
+        return false
+    }
+
+    /**
+     * <i><b>Experimental</b></i>
+     *
+     * This uses lastAttackTime,lastAttackItem variables to manage weapon attack cool-downs.
+     * This assumes that the weapon you just attacked with needs to cool-down before you can attack again,
+     * even if that next attack is with a different weapon.
+     *
+     * Note: This currently only handles melee weapons
+     *
+     * @returns {Promise<void>}
+     */
+    async waitForWeaponCoolDown() {
+        const attackItem = this.lastAttackItem
+        let ticksSinceAttack = (Date.now()-this.lastAttackTime)/(1000*0.05)
+        let weaponTicks = 5 // wait 0.25 sec for attacking with anything else in your hand
+        if (attackItem) {
+            let itemName = attackItem.name.toLowerCase()
+            if (itemName.includes("sword")) {
+                // wait 0.6 second between attacks for swords for 100% damage
+                weaponTicks = 12
+            } else if (itemName.includes("trident")) {
+                weaponTicks = 18 // wait 0.9 second between attacks for tridents for 100% damage
+            } else if (itemName.endsWith("_axe")) {
+                //axes speed actually varies by type...
+                if (itemName.startsWith("wooden") || itemName.startsWith("stone")) {
+                    weaponTicks = 25 // wait 1.25 second between attacks for wooden/stone axe for 100% damage
+                } else if (itemName.startsWith("iron")) {
+                    weaponTicks = 22 // wait 1.1 second between attacks for iron axe for 100% damage
+                } else {
+                    weaponTicks = 20 // wait 1 second between attacks for gold/diamond/netherite axe for 100% damage
+                }
+            } else if (itemName.endsWith("_pickaxe")) {
+                weaponTicks = 17 // wait 0.85 second between attacks for pickaxe for 100% damage
+            }
+        }
```

Comment:
Nice, glad this is finally in here. Eventually we may need to find a way to force players to follow this...